proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,513 @@
+      HasCallStackBacktraceMech
+      deriving (Eq, Show)
+
+    enabledBacktraceMechanisms :: IORef [BacktraceMechanism]
+    enabledBacktraceMechanisms = unsafePerformIO $ newIORef []
+    {-# NOINLINE enabledBacktraceMechanisms #-}
+
+    -- | Set how 'Control.Exception.throwIO', et al. collect backtraces.
+    setEnabledBacktraceMechanisms :: [BacktraceMechanism] -> IO ()
+    setEnabledBacktraceMechanisms = writeIORef currentBacktraceMechanisms
+
+    -- | Returns the currently selected 'BacktraceMechanism's.
+    getEnabledBacktraceMechanisms :: IO [BacktraceMechanism]
+    getEnabledBacktraceMechanisms = readIORef currentBacktraceMechanisms
+
+A ``collectBacktrace`` primitive used by ``throwWithContext``
+simply dispatches to the currently-selected ``BacktraceMechanism``\ s: ::
+
+    module GHC.Exception.Backtrace where
+
+    -- | Collect an 'ExceptionContext' containing backtraces from all enabled
+    -- 'BacktraceMechanism's.
+    collectBacktraces :: HasCallStack => IO ExceptionContext
+    collectBacktraces = do
+        mechs <- enabledBacktraceMechanisms
+        mconcat `fmap` mapM collectBacktrace mechs
+
+    -- | Collect a 'Backtrace' via the given 'BacktraceMechanism'.
+    collectBacktraces' :: HasCallStack => BacktraceMechanism -> IO ExceptionContext
+
+
+    module GHC.Exception where
+
+    -- | Throw an exception. Exceptions may be thrown from purely
+    -- functional code, but may only be caught within the 'IO' monad.
+    -- Backtraces are collected using the backtrace mechanisms selected by
+    -- 'GHC.Exception.Backtrace.enabledBacktraceMechanisms'.
+    throwWithContext :: forall e a. (HasCallStack, Exception e)
+                     => e -> ExceptionContext -> a
+    throwWithContext e ctxt = do
+        -- (implementation simplified for clarity)
+        backtraces <- collectBacktraces
+        raise# (addContext (ctxt <> backtraces) (toException e))
+
+Note that in order to provide ``HasCallStack`` backtraces we propose that a
+``HasCallStack`` constraint be added to ``throw``, ``throwIO``, and similar
+functions. Our prototype implementation suggests that this likely does not
+carry a significant performance impact.
+
+Since some users may want to explicitly opt out of backtrace collection when
+throwing certain exceptions, we also propose to add non-backtrace-collecting
+``throw`` variants: ::
+
+    throwNoBacktrace   :: forall e a. (Exception e) => e -> a
+    throwIONoBacktrace :: forall e a. (Exception e) => e -> a
+
+    throwWithContextNoBacktrace   :: forall e a. (Exception e) => e -> ExceptionContext -> a
+    throwIOWithContextNoBacktrace :: forall e a. (Exception e) => e -> ExceptionContext -> a
+
+
+Examples
+--------
+
+User programs would typically call ``setEnabledBacktraceMechanisms`` during
+start-up to select a backtrace mechanism appropriate to their usage: ::
+
+    main :: IO ()
+    main = do
+        setEnabledBacktraceMechanisms [HasCallStackBacktrace, ExecutionStackBacktrace]
```

Comment:
It means that a `HasCallStack` annotation will not be added to thrown exceptions. In practice I can think of few reasons why one would want to do this since you have already opted in to paying the cost of collecting the `HasCallStack` backtrace when writing the program.

Reply:
🤔 yeah, I would imagine wanting that to be enabled by default